cedar-policy-core/src/parser/err.rs

```diff
@@ -14,51 +14,74 @@
  * limitations under the License.
  */
 
+#![allow(deprecated)] // for `MultipleParseErrors`
+
+use std::fmt::{self, Display, Write};
+use std::iter;
+use std::ops::{Deref, DerefMut};
+
 use lalrpop_util as lalr;
-use std::fmt::Display;
+use miette::{Diagnostic, LabeledSpan, Severity, SourceCode};
+use phf::phf_map;
 use thiserror::Error;
 
+use crate::ast::RestrictedExpressionError;
+
+use crate::parser::fmt::join_with_conjunction;
+use crate::parser::node::ASTNode;
+
+pub(crate) type RawLocation = usize;
+pub(crate) type RawToken<'a> = lalr::lexer::Token<'a>;
+pub(crate) type RawUserError = ASTNode<String>;
+
+pub(crate) type RawParseError<'a> = lalr::ParseError<RawLocation, RawToken<'a>, RawUserError>;
+pub(crate) type RawErrorRecovery<'a> = lalr::ErrorRecovery<RawLocation, RawToken<'a>, RawUserError>;
+
+type OwnedRawParseError = lalr::ParseError<RawLocation, String, RawUserError>;
+
 /// For errors during parsing
-#[derive(Debug, Error, PartialEq, Clone)]
+#[derive(Clone, Debug, Diagnostic, Error, PartialEq)]
 pub enum ParseError {
-    /// Error from the lalrpop parser, no additional information
-    #[error("{0}")]
-    ToCST(String),
-    /// Error in the cst -> ast transform, mostly well-formedness issues
+    /// Error from the CST parser
+    #[error(transparent)]
+    #[diagnostic(transparent)]
+    ToCST(#[from] ToCSTError),
+    /// Error in the CST -> AST transform, mostly well-formedness issues
     #[error("poorly formed: {0}")]
+    #[diagnostic(code(cedar_policy_core::parser::to_ast_err))]
     ToAST(String),
     /// (Potentially) multiple errors. This variant includes a "context" for
     /// what we were trying to do when we encountered these errors
-    #[error("error while {context}: {}", MultipleParseErrors(errs))]
+    #[error("error while {context}")]
     WithContext {
         /// What we were trying to do
         context: String,
         /// Error(s) we encountered while doing it
-        errs: Vec<ParseError>,
+        #[source]
+        #[diagnostic_source]
+        errs: ParseErrors,
```

Comment:
will this hide the actual parse error and only show "error while parsing an EUID" or similar?  Shouldn't the `#[error]` include `{errs}`?

Reply:
The `#[source]` and `#[diagnostic_source]` attributes will cause the inner `ParseErrors` error to be printed out when formatting the `WithContext`. It'll look like:

```
error while doing a thing

caused by:
  multiple parse errors:
    some parse error
    some other parse error
```

Reply:
Notable, if there's only a single parse error in `errs`, it'll look like:

```
error while doing a thing

caused by:
  some parse error
```

since I set up `ParseErrors` to pass everything through transparently if there's only one error in the collection.